nvflare/app_common/abstract/learner2.py

```diff
@@ -0,0 +1,149 @@
+        self.job_id = None
+        self.app_root = None
+        self.abort_signal = None
+        self.current_round = 0
+        self.total_rounds = 0
+
+    def is_aborted(self) -> bool:
+        return self.abort_signal and self.abort_signal.triggered
+
+    def get_shareable_header(self, key: str, default=None):
+        if not self.shareable:
+            return default
+        return self.shareable.get_header(key, default)
+
+    def get_context_prop(self, key: str, default=None):
+        if not self.fl_ctx:
+            return default
+        assert isinstance(self.fl_ctx, FLContext)
+        return self.fl_ctx.get_prop(key, default)
+
+    def debug(self, msg: str):
+        self.log_debug(self.fl_ctx, msg)
+
+    def info(self, msg: str):
+        self.log_info(self.fl_ctx, msg)
+
+    def error(self, msg: str):
+        self.log_error(self.fl_ctx, msg)
+
+    def warning(self, msg: str):
+        self.log_warning(self.fl_ctx, msg)
+
+    def exception(self, msg: str):
+        self.log_exception(self.fl_ctx, msg)
+
+    def critical(self, msg: str):
+        self.log_critical(self.fl_ctx, msg)
+
+    def panic(self, reason: str):
+        self.system_panic(reason, self.fl_ctx)
+
+    def initialize(self, parts: dict):
+        """Initialize the Learner object. This is called before the Learner can train or validate.
+
+        This is called only once.
+
+        Args:
+            parts: components to be used by the Trainer
+        """
+        pass
+
+    @abstractmethod
+    def train(self, dxo: DXO) -> Union[str, DXO]:
+        """Called to perform training. Can be called many times during the lifetime of the Learner.
+
+        Args:
+            dxo: the training input data (e.g. model weights)
+
+        Returns: train result as a DXO if successful; or return code as str if failed.
+
+        """
+        pass
+
+    def get_model_for_validation(self, model_name: str) -> Union[str, DXO]:
+        """Called to return the trained model from the Learner.
+
+        Args:
+            model_name: type of the model for validation
+
+        Returns: trained model for validation
+
+        """
+        pass
+
+    def validate_before_train(self, dxo: DXO) -> Union[str, DXO]:
+        """Validate the current model with the specified weights in dxo before training.
+
+        Args:
+            dxo: the DXO object that contains model weights
+
+        Returns: validation metrics in DXO if successful; or return code if failed.
+
+        """
+        pass
+
+    def validate(self, dxo: DXO, validate_type: str, model_owner: str) -> Union[str, DXO]:
+        """Validate the model with the specified weights in dxo
+
+        Args:
+            dxo: the DXO object that contains model weights
+            validate_type: type of validation
+            model_owner: the owner of the model weights
+
+        Returns: validation metrics in DXO if successful; or return code if failed.
+
+        """
+        pass
+
+    def abort(self):
```

Comment:
No. The abort() method is to be implemented by a a subclass of Learner2; whereas panic() is implemented by the base class, and is to be called by the subclass to end the job in its logic.

Reply:
We'll just have to document it clearly, or use better names for these?

Reply:
So the default implementation for abort() would be calling self.system_panic()? Sounds interesting.

But there is still a difference between the intended abort (server asks the client to abort) and the case that the client couldn't continue due to its own logic. In the first case, it's not an error; whereas the second case is usually a serious error. If we use the "panic" as the default impl, all the log messages produced by panic could be very confusing for the intended abort.

Reply:
Yea, we certainly could improve the names of the two methods. Another reason why they couldn't be combined into one method is the thread issue. In case of "system ask client to abort job", it happens in a separate thread than the training thread; whereas the case of "user intend to abort job" happens in the training thread. Some system/learner uses "training_thread.join()" when aborting the training. If this is called from the training thread itself, it would be a programming error and cause an exception.

Reply:
cancel_job() is exposed to the learner writer so that the training could be aborted gracefully (if the underlying training framework supports graceful abort). If not, then the training process will have to be forcefully killed. A real example of this is ignite-based MONAI training. Before Ignite modified its code to support graceful abort, the training process could only be stopped by killing it. Now Ignite supports abort, the cancel_job() method could now be implemented as "ignite_engine.abort()".